Xray insights batch 1

sbelak opened the pull request:
PR includes end-to-end implementation of the first batch of insights and supporting changes/refactors elsewhere.

Included are:
- [x] Normal range
- [x] Nils 
- [x] Autocorrelation
- [x] Variation range
- [x] Noisiness
- [x] Spikes/dropoffs (outliers)
- [x] Stationarity check
- [x] Trend
- [ ] Composite trend
- [x] Structural breaks
- [x] Multi modality
- [x] Seasonal component
- [x] Zeros

Examples can now be reached (assuming you generate sample data) at:
- nils: x-ray field Orders/discount
- zeros:  x-ray field Orders/discount
- normal-range: x-ray field Orders/Total
- autocorrelation: x-ray card Orders, Sum of Total, Grouped by Created At (month)
- noisiness: x-ray card Orders, Sum of Total, Grouped by Created At (month)
- variation trend: x-ray card Orders, Sum of Total, Grouped by Created At (month)
- seasonality: x-ray card Orders, Average of Quantity, Grouped by Created At (month)
- multimodality: x-ray field Orders/Quantity
- structural breaks: x-ray card 
- outliers: x-ray field Orders/Quantity
- stationary: 
- trend: x-ray card Orders, Sum of Total, Grouped by Created At (month)

sbelak committed:
Add a couple of mock data stories in 2 formats

sbelak committed:
typo

sbelak committed:
Merge branch 'smarter-xray-comparisons-page' of github.com:metabase/metabase into data-stories

sbelak committed:
Add missing arg

attekei committed:
Initial insights frontend

sbelak committed:
added examples with filters and tooltips

attekei committed:
Merge branch 'master' of https://github.com/metabase/metabase into data-stories

attekei committed:
Add example insights to the frontend

attekei committed:
Visual formatting, add link to term definition

attekei committed:
Update regime change insight, have equal height grid

attekei committed:
Extract TermWithDefinition to a reusable module

attekei committed:
Check in integration tests that insight cards are displayed as expected

attekei committed:
Make frontend responsible for storing term definitions and links

sbelak committed:
Refactor out timeseries, math, and insights

sbelak committed:
update tests

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Fix test

sbelak committed:
Add Insight cards

sbelak committed:
Fix NormalRangeinsight

sbelak committed:
Fix nameclash

kdoh committed:
type annotations

kdoh committed:
text and card visual updates

sbelak committed:
Update sample data

sbelak committed:
Merge branch 'xray-insights-batch1' of github.com:metabase/metabase into xray-insights-batch1

sbelak commented:
Examples can now be reached at
gap: Xray field Orders/discount
normal-range: Xray field Orders/Total
autocorrelation: Xray timeseries created_at(by day) x avg(quantity)
noisiness: Xray timeseries created_at(by day) x avg(quantity)

sbelak committed:
remove not null constraint

sbelak committed:
fix variation-trend

kdoh committed:
more text spacing. add feedback widget

kdoh committed:
add translation tags

sbelak committed:
More sound variance trend detection.

sbelak committed:
Merge branch 'xray-insights-batch1' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
update deps

sbelak committed:
Remove incanter dep in favour of distributions.

sbelak committed:
Add seasonality insight, update sample dataset.

sbelak committed:
Cap max-lag at period length for well formed time series

sbelak committed:
More robust async cancelation test

kdoh committed:
instrument feedback

kdoh committed:
tweak insight icon types

kdoh committed:
tweak variation copy

kdoh committed:
ensure cards have the same height

kdoh committed:
add ly

sbelak committed:
Catch async errors

sbelak committed:
More div by 0 checks

sbelak committed:
Merge branch 'xray-insights-batch1' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Correctly handle case where slope=0

sbelak committed:
Add zeros insight. Add docstrings to insights. Minor refactor

sbelak committed:
push iqr into features

sbelak committed:
Add multimodality insight

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Fix failing tests

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Temporariliy disable multimodal insight (to get the PR green)

sbelak committed:
Remove mock insights

sbelak committed:
Fix failing tests due to changes in sample dataset

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

attekei committed:
Show nicely formatted field name in insights, fix an issue with empty stats

sbelak committed:
Add tests, adaptive cutoffs, minor refactor, share y label unit

sbelak committed:
Merge branch 'xray-insights-batch1' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
fix name collision

sbelak committed:
Trigger ci rerun

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Add structural breaks and outliers insights

kdoh committed:
fix multi row card height

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Reenable multimodal insight, list out breakpoints

sbelak committed:
Merge branch 'xray-insights-batch1' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Remove outlier insight from card xrays

sbelak committed:
add stationary insight

sbelak commented:
I'm gonna punt on composite trend in this PR and spend some time implementing https://arxiv.org/pdf/1505.04305.pdf as a comprehensive solution for seasonality, structural breaks, and trends.

sbelak committed:
More comments, more tests, minor refactor

sbelak committed:
Add breaks test

sbelak committed:
Minor touchups

sbelak committed:
comment out middleware tests that fail intermitently

sbelak committed:
Remove unneeded sleep from test

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
fix integration tests

sbelak committed:
Make breaks list pretty.

sbelak committed:
increase timeout for xray tests

sbelak committed:
Fix insigths integration test, up timeouts

sbelak committed:
Fix insights test

sbelak committed:
Speed up variation-trend and stationarity insight

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Add trend insight

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Reenamble middleware test

sbelak committed:
Optimize some insights using transducer magic

sbelak committed:
minor refactor

sbelak committed:
disable troublesome middleware tests

sbelak committed:
minor refactor

sbelak committed:
Update to latest kixi.stats

sbelak committed:
Add regularization to trend fitting and correctly handle NaNs

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class AutocorrelationInsight extends Component {
+    static insightType = "autocorrelation"
+    static title = "Autocorrelation"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class NoisinessInsight extends Component {
+    static insightType = "noisiness"
+    static title = "Noisy data"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class ZerosInsight extends Component {
+    static insightType = "zeros"
+    static title = "Zeros in your data"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+        return (
+            <InsightText>
+                Your data has a { quality } <TermWithDefinition definition={autocorrelationDefinition} link={autocorrelationLink}>
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+const variationTrendDefinition = "How variance in your data is changing over time."
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class VariationTrendInsight extends Component {
+    static insightType = "variation-trend"
+    static title = `Trending variation`
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+        return (
+            <InsightText>
+                Looks like this data has grown { mode }ly  <TermWithDefinition definition={variationTrendDefinition} link={varianceLink}>
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class SeasonalityInsight extends Component {
+    static insightType = "seasonality"
+    static title = "Seasonality"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+        return (
+            <InsightText>
+                Your data has a { quality } seasonal compoment.
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+const multimodalDefinition = "Data distribution with multiple peaks (modes)."
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+    render() {
+        return (
+            <InsightText>
+                Your data looks to be <TermWithDefinition definition={multimodalDefinition} link={multimodalLink}>
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class OutliersInsight extends Component {
+    static insightType = "outliers"
+    static title = "Outliers"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class StructuralBreaksInsight extends Component {
+    static insightType = "structural-breaks"
+    static title = "Structural breaks"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+const stationaryDefinition = "Mean does not change with time."
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+        return (
+            <InsightText>
+		It looks like your data has
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+	return (
+		<InsightText>
+                Your data looks to be <TermWithDefinition definition={stationaryDefinition} link={stationaryLink}>
```
`t`

camsaul replied:
this indentation looks weird

shouldn't it be

```javascript
        return (
            <InsightText>
                Your data looks to be <TermWithDefinition definition={stationaryDefinition} link={stationaryLink}>
```

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+export class TrendInsight extends Component {
+    static insightType = "trend"
+    static title = "Trend"
```
`t`

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+/*
```
I'd like to remove this commented out insight if we're not shipping it as part of this batch.

kdoh commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+
+	return(
+		<InsightText>
+		Your data seems to be { mode } { shape }.
```
`t`

kdoh commented on frontend/src/metabase/xray/components/Insights.jsx:
```diff
@@ -0,0 +1,27 @@
+        const { features } = this.props;
+
+        const parametrizedInsights = Object.entries(features["insights"])
+            // temporary hacks as we have two formats
```
Can we elaborate on this? Examples of the formats maybe or some reason about why there are two?

attekei replied:
whoops. that's an old hack which can be safely removed

attekei replied:
(that was from era when we were trying to decide whether the backend provides complete sentences or just the computed values, we decided to go with the latter approach)

sbelak replied:
removed

kdoh commented on frontend/src/metabase/xray/containers/CardXRay.jsx:
```diff
@@ -96,6 +97,12 @@ class CardXRay extends Component {
                         </div>
+                        { xray.features["insights"] &&
+                            <div className="mt4">
+                                <Heading heading="Takeaways" />
```
`t`

kdoh commented on frontend/src/metabase/xray/containers/FieldXray.jsx:
```diff
@@ -135,6 +137,12 @@ class FieldXRay extends Component {
                             </div>
+                            { features["insights"] &&
+                                <div className="mt4">
+                                    <Heading heading="Takeaways" />
```
`t`

kdoh commented on frontend/test/parameters/parameters.integ.spec.js:
```diff
@@ -60,8 +60,8 @@ const getRelativeUrlWithoutHash = (url) =>
 const COUNT_ALL = "200";
-const COUNT_DOOHICKEY = "56";
-const COUNT_GADGET = "43";
+const COUNT_DOOHICKEY = "51";
```
I'm assuming these values changed cause of the update to the Sample Dataset? Is there any way to avoid this being so fragile and causing test failures?

attekei replied:
yep, they did. we could simply change the lines where those variables are used to be less specific; they could verify that the scalar has a value of >= 0 instead of checking the exact value

sbelak replied:
I've started doing that, but the problem soon becomes that sometimes (eg. when testing binning) you want to test for particular values. Some can probably be worked around by changing queries. 

The current approach is not ideal especially as the sample dataset will probably change more in the coming months, but fixing this probably deserves it's own PR as it's not a trivial change.

kdoh commented on frontend/test/xray/xray.integ.spec.js:
```diff
@@ -332,7 +347,7 @@ describe("xray integration tests", () => {
             app = mount(store.getAppContainer())
 
-            await store.waitForActions([LOAD_CURRENT_USER, INITIALIZE_SETTINGS])
+            await store.waitForActions([LOAD_CURRENT_USER, INITIALIZE_SETTINGS], { timeout: 20000 })
```
The timeout seems pretty long here. Might be worth commenting on?

attekei replied:
I think that the timeout is just unnecessary cruft caused by a copy-paste in this case

kdoh reviewed (changes requested):
Looks like there's a bit of cleanup to do in the InsightCard component. I can take care of that if @attekei is busy.

senior commented on src/metabase/feature_extraction/timeseries.clj:
```diff
@@ -0,0 +1,174 @@
+  ([period ts]
+   (decompose period {} ts))
+  ([period opts ts]
+   (when-let [period (period-length period)]
```
I'm having difficulty matching up the tests (and the boolean return value) with the code here. It looks like it's config driving the setup of the StlConfig objects, but what happens after that?

sbelak replied:
It does the actual decomposition in the cleanup airity of the transducer. Clearly too cute. Refactored.

senior commented on test/metabase/feature_extraction/math_test.clj:
```diff
@@ -0,0 +1,91 @@
+(expect
+  (approximately 5.5 0.1)
+  (transduce identity magnitude [1 2 3 4]))
+(expect
```
spacing issues here

camsaul replied:
bump

sbelak replied:
This one is intentionally missing an empty line as it would have been just one expect if it weren't for the approximate which doesn't play nicely with vectors.

camsaul replied:
can we add a comment saying as much? That way people know why it's done that way

sbelak replied:
Sure. Also, I'm not super attached to this, just seemed like the right thing to do at the time.

sbelak replied:
note added

senior commented on src/metabase/feature_extraction/timeseries.clj:
```diff
@@ -0,0 +1,174 @@
+  "Quarter-of-year functionality"
+  (quarter [dt] "Return which quarter (1-4) given date-like object falls into."))
+
+(extend-type java.util.Date
```
extend protocol would save some space, just a style suggestion though

senior commented on src/metabase/feature_extraction/insights.clj:
```diff
@@ -0,0 +1,225 @@
+(ns metabase.feature-extraction.insights
```
@sbelak I couldn't find the tests covering this ns, could you point me to them?

camsaul replied:
+1

sbelak replied:
added here https://github.com/metabase/metabase/pull/6266/files#diff-b4d18e3f6af9ee0f6fc3ca4160d8f254

camsaul replied:
Oh ok I see tests have been added at metabase.feature-extraction.insights-tests

senior reviewed (commented):
Generated the uberjar from master and compared it with the one here, the new jars add roughly 7 MB.

senior commented on test/metabase/middleware_test.clj:
```diff
@@ -220,7 +220,7 @@
           [first-second second-second (string/trim eventually)])))))
 
-
+(comment
```
This makes me a bit nervous as it happened very infrequently before, and it sounds like now it's pretty frequent. If we can convince ourselves this is a test only problem (the tests do look a little troublesome anyway) and this is a feature we care about, we should write up an issue for fixing it, note it here and work on it soon.

camsaul replied:
have we figured out what to do with this yet?

sbelak replied:
Not really. Just eliminated one option. Here's @senior's take after we went through:

"One of the things I was worried about with the async processing going on with xrays and the streaming test failures we were seeing is overloading the clojure threadpool that is used whenever you invoke a future in clojure. I was thinking that pool was bounded, but I was confused, it's the agent pool that is bounded, not the pool that futures use.

After chatting some with Simon, I don't think this is the cause of our issue, but something to keep in mind is where our background/async tasks are running and where the back pressure is. In the case of sending async emails or the xray code, it's running with the unbounded clojure "soloExecutor" threadpool."

salsakran commented:
What is causing the increase in jar size by 7 megs?

That's a big enough increase, that I'd like something pretty compelling to be the reason.

camsaul commented on frontend/src/metabase/lib/formatting.js:
```diff
@@ -361,6 +361,11 @@ export function slugify(name: string) {
+// Adds commas and spaces between array items and replaces the comma between two last items with "and"
+export function formatListOfItems(items: []): string {
+    return [items.slice(0, -1).join(', '), items.slice(-1)[0]].join(items.length < 2 ? '' : ' and ');
+}
```
since all the code that's using this is commented out we may as well just remove the function for now and save it for a PR that actually uses that

failing that, we should write tests for it. I don't see any tests :eyes:

attekei replied:
yup we can remove this function

camsaul replied:
+1

sbelak replied:
removed

camsaul commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+        // construct the question with filter
+        return (
+            <InsightText>
+                You have { quality } missing (null) values in your data.
```
👍  if we're in the process of doing i18n we should definitely make sure all new PRs going forward have everything tagged up so we keep making forward progress instead of getting further away from our goal of 100% of things tagged

attekei replied:
yeah, this should be one of the standard requirements for merging a PR in addition to tests (and instrumentation in case of bigger features like this)

camsaul commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+    static icon = "warning"
+
+    render() {
+        const { quality, filter, features: { table } } = this.props
```
@kdoh what's our policy on semicolons in Javascript? 

Personally I don't like implicit semicolons coming back to bite me in the ass so I've always preferred living in a land where we are explicit and have the linter prevent implicit ones. That way we don't make mistakes like the infamous

```javascript
return
{
    a: 100
}
```

Which JS interprets as

```javascript
return;
{
    a: 100
};
```

or 

```javascript
a = b + c
(d + e).print()
```

which gets interpreted as 

```javascript
a = b + c(d + e).print();
```

I know it's fun to be a JSHipster and pretend Javascript is just like python and doesn't need semicolons but the fact is sooner or later omitting them is going to introduce bugs. I think [this commenter on HN](https://news.ycombinator.com/item?id=2183842) says it better than I ever could:

> Omitting semicolons in JavaScript is like omitting braces from one line if blocks.
It's all well and good until someone turns this:
>
> ```c
> if (conditional)
>     statement;
> ````
>
> Into this without realizing what they've just done:
>
> ````c
> if (conditional)
>     statement;
>     statement2;
> ````
>
>Yes, you can skip semicolons. And yes it will work the vast majority of the time. But the amount of time you save will be eclipsed by that one time someone makes a change to the code that breaks in an unpredictable subtle way due to your purely aesthetic shorthand.
>Writing maintainable code is about learning humility. There's no need to show off your understanding of the deeper points of JavaScript semicolon insertion in your production code base. Just insert the semicolons, I promise you you will learn to not be distracted in no time.

kdoh replied:
So we don't have a current lint rule or enforcement of this which I know is goofy and no good. I personally tend to be on team no semi-colons because I think the overall noise reduction makes it easier for me personally to write JS and I know the gotchas so I can avoid them, but I understand the arguments against it.

In my mind, reviewing code for style tends to get into personal tastes and philosophical debates I'd rather continue to not read on HN, so my real answer is that we should have robots do it for us. My vote would be to use https://prettier.io/ (or similar) for style rulse vs a linter so that team members and contributors can just write how they write and know it'll be formatted properly on the other side. Here's [more info](https://prettier.io/docs/en/comparison.html) on that.

My vote would be to use the time it would take to go add `;` everywhere or remove it, to set up said robot. We can add a lint rule to our current lint set up if it's really nagging at you, but it's inconsistent enough throughout the project that in either case we'd have to run some sort of ``--fix` which would create an enormous diff, so in my mind I prefer the solution that gives me more time to think about the what and not the formatting.

attekei replied:
the earlier post-commit hook configuration of Prettier made it cumbersome to use – I would love to have auto-formatting as part of my daily workflow but not sure how it would feel most natural

camsaul commented on frontend/src/metabase/xray/components/InsightCard.jsx:
```diff
@@ -0,0 +1,368 @@
+        return `${mode} ${shape} period from ${formatTimeWithUnit(from, resolution)} until now`
+    }
+    render() {
+        let { breaks } = this.props
```
you can use `const` here

camsaul replied:
`const` in javascript is just like a reference in C++ in the sense that you can't unseat it (although there's no guarantees about it not being `null`). It doesn't guarantee anything about it actually being "readonly" so unless you're planing on reassigning `breaks` below we should use `const`. It's easier to follow code that only uses `let` for things that might be unseated

camsaul commented on frontend/src/metabase/xray/components/StatGroup.jsx:
```diff
@@ -12,14 +12,14 @@ const StatGroup = ({ heading, xray, stats, showDescriptions }) =>
                                 />
                             </li>
-                        )
+                        ) : null
```
Personally I preferred this how it was, it was more readable than the ternary operator if only because there's 7 lines between when it starts and when it ends. Ternaries get hard to read when they span that many lines

attekei replied:
The idea here is that in some instances  `xray[stat]` had a value but `xray[stat].value` was missing, causing the value of `xray[stat]` to leak to html tree because of how `&&` operator functions.

camsaul commented:
@sbelak are you using tabs in your files? We've been purely using spaces for indentation (see the [`.editorconfig` file](https://github.com/metabase/metabase/blob/master/.editorconfig)). Check out the `editorconfig-mode` package for Emacs which will apply that file automatically for you.

At any rate a few of us have our editiors set up to automatically untabbify any file we save so we're going to run into a huge mess of meaningless file changes whenever we touch one of the files from this PR

camsaul commented on project.clj:
```diff
@@ -58,12 +59,14 @@
                  [compojure "1.5.2"]                                  ; HTTP Routing library built on Ring
                  [crypto-random "1.2.0"]                              ; library for generating cryptographically secure random bytes and strings
+                 [distributions "0.1.3-SNAPSHOT"                      ; Probability distribution
+                  :exclusions [org.clojure/clojure]]
```
What @senior said about the `jdistlib` dep applies to this too. Relying on a `-SNAPSHOT` dependency is risky AF because it could change out from under us without notice. Preferable to stick to something like `0.1.2` instead of an unfinalized snapshot

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -60,18 +61,27 @@
 (defn- random-company-name []
   (first (company/names)))
 
+(defn- rejection-sample
```
We don't enforce docstrings for private functions like this, but there's not rule against having them anyway. TBH I'm not a stats expert so the term `rejection-sample` is meaningless to me. It would be great if we even had a short blurb that would help someone playing around in this namespace understand what exactly is going on without having to go ask anyone else

camsaul replied:
`random-price` below this is clear enough but for things that aren't immediately clear to other people who have to work on the code I would prefer we try to at least make the code somewhat documented, private function or no. 

I think fixing the frontend bugs I fixed the last couple weeks took me 3-4x as long as it needed to because it took me so long to understand what the code was supposed to be doing. I think the viz lib averaged maybe one comment per 400 lines of code.

Spending a couple extra minutes writing explanations of what everything that's not immediately obvious does and how it works with without a doubt save us time in the future when we go back to tweak it

camsaul replied:
+1

camsaul replied:
Sorry, this was already fixed

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -194,20 +204,33 @@
+(defn- sometimes
+  [p f]
+  (when (> p (rand))
+    (f)))
```
So... this function sometimes does `f`, and `p` is the probability of doing it? That's not clear at all until you read the actual code for the function. I would vote in favor of a slightly more verbose name that more adequately summarizes what is going on. `do-sometimes-with-probability` or something like that is a lot clearer. At any rate a docstring would have *also* been a good way to clear this up

camsaul replied:
+1

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -60,18 +61,27 @@
-        (+ min))))
+  (let [range    (- max min)
+        m1       (+ min (* range 0.3))
+        m2       (+ min (* range 0.75))
```
What's an `m1` and `m2`? A more verbose name or a comment explaining what these are would make this code a lot easier to understand 

https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#prefer-more-verbose-names

camsaul replied:
+1

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -248,6 +271,43 @@
+  (for [x xs]
+    (update x k * (seasonality-map (season-fn x)))))
+
+(defn add-outliers
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

camsaul replied:
+1

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -248,6 +271,43 @@
+                   (update x k * (+ 1 (* (/ i n) (- (rand 2) 0.9)))))
+                 xs)))
+
+(defn add-seasonality
```
These sorts of things would benefit from having a short little blurb too about what exactly it is they do. One day someone else is going to have to go in and make changes to this code and we will be making their lives a lot easier if we make it easy for them to understand what everything does

camsaul replied:
+1

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -248,6 +271,43 @@
+    (update x k * (seasonality-map (season-fn x)))))
+
+(defn add-outliers
+  [mode n k xs]
```
it's not super clear to me what `mode` means here. It looks like this function gets called exactly twice, once with a mode of `:share`, and once with a mode of `:count`, but only `:mode` actually ever does anything. That's something it would be good to document somewhere

camsaul replied:
+1

camsaul commented on src/metabase/feature_extraction/comparison.clj:
```diff
@@ -1,48 +1,16 @@
-
-(defn head-tails-breaks
-  "Pick out the cluster of N largest elements.
-   https://en.wikipedia.org/wiki/Head/tail_Breaks"
```
These functions for example are 100x easier to understand because of the docstrings. 👍

camsaul reviewed (changes requested):
*  ~We should make sure to i18n tag all new strings~
*  Please don't use tabs in files. Configure your editor to use spaces
*  The `SNAPSHOT` dependencies are dangerous, like @senior said we should pin them down so they can't randomly change and break stuff for us
*  The sample dataset generation code could really use some documentation. At some point in the future someone else is going to have to make changes to it (new columns or other tweaks) and will have trouble understanding what all of the moving parts do.

attekei commented:
@kdoh if you have time for doing the frontend fixes, that would be awesome – I'm really short on time this week – but that said, I can help in fixing individual issues if something feels complicated for you.

sbelak commented on project.clj:
```diff
@@ -74,6 +77,10 @@
                  [medley "0.8.4"]                                     ; lightweight lib of useful functions
                  [metabase/throttle "1.0.1"]                          ; Tools for throttling access to API endpoints and other code pathways
                  [mysql/mysql-connector-java "5.1.39"]                ;  !!! Don't upgrade to 6.0+ yet -- that's Java 8 only !!!
+                 [jdistlib "0.5.0-SNAPSHOT"                           ; Distribution statistic tests
```
We did cut it (it's "mine" on Clojars, as I needed a non JDK 8 version). Will change though.

kdoh committed:
cleanup and translation

sbelak committed:
Implement code review suggestions part 1

sbelak commented:
- Implemented all the clj suggestions (sorry for the mess in generate.clj, forgot to do a cleanup pass and the linter didn't smack me).
- Fixed indentation in jsx (sorry^2) and removed commented out code
- Remove dependency on distributions library. This should bring the jar size down.

Still needs
- [ ] tests for insights that don't just wrap library functions (which are covered by tests)
- [ ] translation strings

sbelak committed:
Make linter happy

sbelak committed:
Make outliers work with empty input

kdoh committed:
merge

kdoh commented:
I added some translation tags and a few other bits of cleanup to `InsightCard.` There are a few [more complex](https://github.com/metabase/metabase/blob/6683a2ebc260842be4235796d9b0e37d7ed9fdbb/frontend/src/metabase/xray/components/InsightCard.jsx#L90) nestings of JSX elements that I'm a little unsure of how to handle however. Could use guidance on those.

sbelak committed:
Added missing tests

sbelak committed:
Make zero and nil insights work with no data

sbelak committed:
Merge branch 'master' of github.com:metabase/metabase into xray-insights-batch1

sbelak committed:
Make variation trend test more robust

sbelak committed:
Make variance trend test (even) more robust

sbelak committed:
Add centroid test

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -12,7 +12,8 @@
              [lorem :as lorem]
              [name :as name]]
-            [incanter.distributions :as dist]
+            [distributions.core :as dist]
```
+1

camsaul commented on sample_dataset/metabase/sample_dataset/generate.clj:
```diff
@@ -6,13 +6,14 @@
              [jdbc :as jdbc]]
             [clojure.math.numeric-tower :as math]
             [clojure.string :as s]
+            [jdistlib.core :as dist]
```
ns sorting

sbelak replied:
fixed
